src/tools/schema2graph.ts

```diff
@@ -40,9 +40,12 @@ export class SchemaSource {
 
 export class SchemaNode {
   constructor(
-    readonly schema: Schema,
+    // Schemas can be null when the node represents a type variable with no constraints.
+    readonly schema: Schema | null,
     readonly particleSpec: ParticleSpec,
-    readonly allSchemaNodes: SchemaNode[]
+    readonly allSchemaNodes: SchemaNode[],
+    // Type variable associated with the schema node.
+    readonly fromVariable: string | null = null
```

Comment:
I think what happened is that this used to be a boolean, and then you found you need to have the name as well. So the type change but the name doesn't fit great IMO. Maybe `sourceVariable` or `variableName` would be better now?